application/common/models/User.php

```diff
@@ -40,6 +48,195 @@ public function rules()
     }
 
     /**
+     * Find or create local user. Fetch/update user data from personnel.
+     * @param string $username
+     * @return User
+     * @throws NotFoundHttpException
+     * @throws \Exception
+     * @throws NotFoundException
+     */
+    public static function findOrCreate($username)
+    {
+        /*
+         * If username looks like an email address, search by email
+         */
+        /** @var PersonnelUser $personnelUser */
+        if (substr_count($username, '@') > 0) {
+            $personnelUser = \Yii::$app->personnel->findByEmail($username);
+        } else {
+            $personnelUser = \Yii::$app->personnel->findByUsername($username);
+        }
+
+        $user = self::findOne(['employee_id' => $personnelUser->employeeId]);
+        if ( ! $user) {
+            $user = new User();
+            $user->employee_id = $personnelUser->employeeId;
+            $user->first_name = $personnelUser->firstName;
+            $user->last_name = $personnelUser->lastName;
+            $user->idp_username = $personnelUser->username;
+            $user->email = $personnelUser->email;
+            if ( ! $user->save()) {
+                /**
+                 * @todo add logging with model validation errors
+                 */
+                throw new \Exception('Unable to create new user', 1456760294);
+            }
+        } else {
+            $user->updateProfileIfNeeded(
+                $personnelUser->firstName,
+                $personnelUser->lastName,
+                $personnelUser->username,
+                $personnelUser->email
+            );
+        }
+
+        return $user;
+    }
+
+
+    /**
+     * Update local user record if given properties are different than currently stored
+     * @param string $firstName
+     * @param string $lastName
+     * @param string $username
+     * @param string $email
+     * @return bool True if profile was updated, false if not
```

Comment:
It might be worth clarifying that `false` here means it wasn't updated because it didn't need to be updated (it all matched). As it is, someone could misread it as "it tried to update and failed" (such as if a validation rule failed).